Exploring V2 Features and Design

Darshan808 opened the pull request:
I am exploring the V2 features and design of jupyterlab-notify. This PR includes my initial implementation which I built upon existing branch 'Integrate with UI', and I want to confirm whether my approach is correct. Any feedback or guidance would be extremely helpful in refining the implementation.
There are many todo works still left.

https://mybinder.org/v2/gh/Darshan808/jupyterlab-notify/v2-test?urlpath=lab

krassowski committed:
Initial work on settings and icon in cell toolbar

krassowski committed:
More work on toolbar integration

Darshan808 committed:
explore v2 design

Darshan808 commented:
Currently, we're assuming a local SMTP server and auto-fetching email IDs via getUser() method of getpass module. How should we proceed now? I propose we ask users for their SMTP settings (server, port, username, password) so they can use third-party services. For Slack notifications, we could ask for the Slack bot token and the channel/user_id. Any suggestions on how best to collect this info (e.g., config file, interactive setup, etc.)?

Darshan808 committed:
add notification feature

Darshan808 committed:
add email and slack options

Darshan808 commented:
Hi @krassowski 

Quick questions:

- UI for Notification Channels:
We’re using bell icons for notification settings, but I’m thinking of adding a dropdown in the cell-toolbar to let users choose between Gmail, Slack, both, or none. Does that placement sound good, or do you have another suggestion?

- Config Setup:
Right now, I'm assuming users will set up their Slack bot and provide the necessary details (token, user/channel ID, email) in ~/.jupyter/jupyterlab_notify_config.json. Is that a good approach, or should we consider something else for storing this info?

Thanks for the feedback!

krassowski commented:
> choose between Gmail, Slack, both, or none. Does that placement sound good, or do you have another suggestion

I would think that this could just go to settings for now? My reasoning is that users will more often toggle if they want notifications, less often how long the threshold should be and rarely configure the way they want them delivered so the UI should reflect this.

If we get a request to include that in the dropdown then we can explore it but for now I would keep it simple.

> users will set up their Slack bot and provide the necessary details (token, user/channel ID, email) in ~/.jupyter/jupyterlab_notify_config.json

Good question. I think we do not want to make the token readable in UI so the standard settings would not be a good place for it. However, we may want to allow to set the token in the UI - this is how jupyter-ai does it (you can set a key from UI and see the key is set, but not read its value). That said, I think a config file without UI is the right approach for now so yes `jupyterlab_notify_config.json` is the way!

krassowski commented:
Maybe there should be some rules in the settings UI to set like "If it takes more than 30 minutes notify via email". If it takes more than 10 minutes notify via Slack". I will ask around for what we should do here.

I would imagine a future version could even have "notify via email and slack if the browser tab is not open" but I think this is a non-goal for now.

Darshan808 committed:
add settings and smtp config

Darshan808 commented:
@krassowski 
I see `global timeout` and `custom timeout`. I don't understand why are they different.
Since users are able to change global timeout value, I think both can be merged into a single timout.

Darshan808 committed:
add-metadata-to-new-cells

krassowski commented:
So the idea is that a user could define a "global" timeout for all cells across all notebooks, say 5 minutes. But then, when they work with a single cell which takes on average 4 minutes to execute, they want to change it for that single cell (without affecting timeout of other cells) - so they would use a "custom" timeout for that cell.

Maybe a different code name would be better, feel free to propose something.=

Darshan808 committed:
add-threshold-notification

Darshan808 commented:
Most of the discussed features have been implemented. Now, the remaining tasks are code cleanup and adding tests. I expect to have it ready for review within a week.

Darshan808 commented:
For email/Slack notifications, we currently have the `cell ID`, `document ID`, `execution status` (success/failure), and `kernel error` (if any). We could include the `cell ID` and `document ID` in the notification, but I don’t think that would be very user-friendly. Do you have any suggestions on how we can make it more intuitive for users to identify which cell triggered the notification?

Darshan808 committed:
improve-frontend

Darshan808 committed:
improve-backend

Darshan808 commented:
@krassowski

I believe the implementation is now ready for review, as I’ve integrated most of the features. While you review it, I’ll focus on writing tests.

Regarding testing the frontend, can I use Galata/Playwright, or should I stick with Jest for the frontend tests?

Additionally, I noticed that pre-commit hooks and other workflows haven’t been configured yet. I’ll also take care of setting those up.

Darshan808 commented:
> they want to change it for that single cell

I’ve been considering the timeout implementation. We initially decided not to let users customize each cell’s timeout individually. Right now, we have global and custom timeouts, and users can choose between the two. However, with the current setup, it seems almost the same as allowing users to change the timeout value and then run the cell, rather than having the two timeout options.

What are your thoughts on this? How should we proceed with the custom timeout feature? And how should we allow users to customize timeouts for each cell?

Darshan808 committed:
add-server-tests

Darshan808 committed:
add-test-dependencies

Darshan808 commented:
I have one more question: Currently, we are using REST APIs to register new cells for notifications. Since the server handles email and Slack notifications while the frontend manages desktop notifications, I am maintaining two separate lists of cells registered for notifications—one on the frontend and one on the server.

However, I believe that using a socket connection instead of REST APIs would make the code more DRY, as we could maintain a single list of registered cells on the server and emit events to the frontend for desktop notifications.

What do you think about this approach?

Darshan808 commented:
One more thought while writing the tests
If the user has set the notification mode to `global` or `custom` timeout, the notification will trigger when the specified timeout is reached. However, in cases where the cell execution finishes before the timeout, should we still trigger a notification immediately, or should the user only be notified when the timeout occurs, regardless of early execution completion?

Darshan808 committed:
add-playwright-tests

Darshan808 commented:
To change settings of a plugin, I am currently doing
```javascript
  // Enable email in settings
  await page.evaluate(async () => {
    await window.jupyterapp.serviceManager.settings.save(
      'jupyterlab-notify:plugin',
      JSON.stringify({mail: true})
    );
  });
  
  // Reload so that new setting is applied
  await page.reload()
```
Is there a simpler or more efficient way to do this?

Darshan808 commented:
I am encountering a bug after installing `jupyter-server-nbmodel`. When running any Playwright test that involves executing cells, I always get a "Cell execution Failed" notification, even though the cells execute correctly. 
To debug, I added some print statements:
```javascript
    NotebookActions.executed.connect((_, args) => {
      console.log(args.cell.model.id, args.success)
      handleNotification(args.cell.model.id, args.success);
    });
``` 
- **Locally:** It outputs `<cell_id> true`, indicating the execution status is correct.  
- **During Testing:** It outputs `<cell_id> false`, even when the cells are actually running fine.  

I have been investigating this but haven’t pinpointed the issue yet. Do you have any idea why this might happen during testing?

Darshan808 committed:
add-workflows-and-pre-commit

Darshan808 committed:
try-fixing-ci

Darshan808 committed:
try-ui-test-fix

Darshan808 committed:
try-fix-build-ci

Darshan808 committed:
update-eslintignore

Darshan808 committed:
modify-ui-test-workflow

Darshan808 committed:
try-fix-ui-workflow

Darshan808 committed:
modify-ui-test-workflow

Darshan808 committed:
try-fix-workflow

Darshan808 commented:
Finally! 🎉
Since I could not test these workflows locally, I needed to modify a bit and push in order to check. Sorry for the messy commit history!

Darshan808 committed:
bump: upload-artifact to v4

Darshan808 committed:
Merge branch 'integrate-with-ui' into v2-test

Darshan808 committed:
wait-for-notification

Darshan808 commented:
I needed the changes from `main` in `deshaw/jupyterlab-notify` to allow the CI to pass, so I updated `main` here and rebased the branch `integrate-with-ui` onto `main`. However, some unexpected things happened, and merge conflicts arose. I’ve fixed them now.

The only test that isn't passing in CI is `Check Links`, and it's due to the links in the changelog. Any idea how to fix these?

Darshan808 committed:
update-readme

Darshan808 committed:
add-dummy-token

Darshan808 committed:
fix-readme

Darshan808 committed:
lint-fix

krassowski commented on README.md:
```diff
@@ -5,77 +5,75 @@
-### Register magics
+The `jupyterlab-notify` extension allows you to receive notifications about cell execution results in JupyterLab. Notifications are configured through cell metadata or the JupyterLab interface, providing seamless integration and easier management of notification preferences. Notifications can be sent via desktop pop-ups, Slack messages, or emails, depending on your configuration.
+
+**Important Note**: JupyterLab Notify v2 supports `jupyter-server-nbmodel`(>= v0.1.1a2), enabling notifications to work even after the browser is closed. To enable browser-less notification support, install it with:
```
We can render it nicely with https://github.com/orgs/community/discussions/16925

```suggestion
> [!NOTE]
> JupyterLab Notify v2 supports `jupyter-server-nbmodel`(>= v0.1.1a2), enabling notifications to work even after the browser has been closed. To enable browser-less notification support, install it with:
```

krassowski commented on README.md:
````diff
@@ -5,77 +5,75 @@
-```python
-%load_ext jupyterlab_notify
+```bash
+pip install jupyter-server-nbmodel
````
I think this also requires `"jupyter-docprovider>=1.0.0b1", "jupyter-server-ydoc>=1.0.0b1"` as in the `[lab]` optional group of `jupyter-server-nbmodel` ([here](https://github.com/datalayer/jupyter-server-nbmodel/blob/efc2f849e0f7dd388ac93d0c1b1fca3991885707/pyproject.toml#L32)).

Should we add it to as an optional dependency group? Then we could include the version pin for `jupyter-server-nbmodel` and the other dependencies and here just recommend:

```suggestion
pip install jupyterlab-notify[server-side-execution]
```

Darshan808 replied:
Yes, I think this is better. Can we keep the optional dependencies as just `[server]` ? or we could name it `[server-exec]`

Darshan808 replied:
Since the outputs aren't pushed to the UI without `jupyter-collaboration`, it would be better to keep it as an optional dependency under the `[server]` group.

krassowski replied:
Yes, but we do not want to depend on entire `jupyter-collaboration` but just on `jupyter-docprovider` and `jupyter-server-ydoc`.

As for naming, I think explicit is better than implicit. I would vote for `server-side-execution` but `server-exec` is fine.

krassowski commented on README.md:
```diff
@@ -5,77 +5,75 @@
-### Notify completion of single cell:
+### Configuration
+
+To enable Slack and email notifications, create a configuration file at `~/.jupyter/jupyterlab_notify_config.json` with the following format:
```
jupyter --
```suggestion
To enable Slack and email notifications, create a `jupyterlab_notify_config.json` configuration file under a path listed in `jupyter --paths` config section (e.g. `~/.jupyter/jupyterlab_notify_config.json`) with the following format:
```

krassowski commented on README.md:
````diff
@@ -5,77 +5,75 @@
 ```
 
-### Mail output upon completion (with optional title for successfull execution)
+- `slack_token`: A valid Slack bot token with `chat:write` permissions. Refer [this article](https://help.thebotplatform.com/en/articles/7233667-how-to-create-a-slack-bot) to create your own slack bot
````
Can we point to official slack docs instead, for example this one?

```suggestion
- `slack_token`: A valid Slack bot token with `chat:write` permissions. Refer [this article](https://api.slack.com/quickstart) to create your own slack bot
```

krassowski commented on README.md:
```diff
@@ -5,77 +5,75 @@
 
-![notify-extension-in-action](https://github.com/deshaw/jupyterlab-notify/blob/main/docs/notify-screenshot.png?raw=true)
-
-This is inspired by the notebook version [here](https://github.com/ShopRunner/jupyter-notify).
```
Let's keep it as "The initial version of this extension was inspired by ...", probably at the bottom of README.

krassowski commented on README.md:
```diff
@@ -5,77 +5,75 @@
-### Disable notifications
+1. Go to Settings Editor.
+2. Select notify.
+3. Set "globalTimeout": 30 (in seconds) to apply to cells using the global-timeout mode.
```
```suggestion
3. Set "globalTimeout": 30 (in seconds) to apply to cells using the `global-timeout` mode.
```

krassowski commented on schema/plugin.json:
```diff
@@ -12,6 +12,19 @@
+        "global-timeout",
+        "custom-timeout"
+      ],
+      "default": "never"
```
Should it default to `global-timeout`?

krassowski commented on pyproject.toml:
```diff
@@ -22,10 +22,23 @@ classifiers = [
+    "pytest-mock",
+    "jinja2",
+    "slack-sdk",
+    "pydantic",
```
It's already in `dependencies`

```suggestion
```

krassowski commented on pyproject.toml:
```diff
@@ -22,10 +22,23 @@ classifiers = [
 dependencies = [
-    "jupyter_server>=2.0.1,<3"
+    "jupyter_server>=2.0.1,<3",
+    "pydantic>=2.0"
```
It looks like this is used only for `NotificationParams`. Could it be a `dataclass` instead? This is because `pydantic` in past broke many projects with major release causing a lot of maintenance headaches, so it if it is only used for a single class then maybe it is not worth it.

krassowski commented on style/icons/bell-clock.svg:
This icon has a noticeable different outline size for compared to the other icons

![image](https://github.com/user-attachments/assets/3cbe8673-d028-4f80-a42e-bc856887858a)

Compared with:

![image](https://github.com/user-attachments/assets/f0a9fd83-766a-4028-8acd-11ff89775e93)

![image](https://github.com/user-attachments/assets/d141a6bc-ffea-4452-a25a-ab33328de538)

Also, it looks like the source could/should be cleaned up and `jp-icon3` class added. I would recommend hand-crafting it on SVG level.

Darshan808 replied:
Redesigned it, Could you verify it it still needs improvement?

krassowski commented on src/index.ts:
```diff
@@ -2,150 +2,414 @@ import {
-    icon: bellOutlineIcon // TODO: custom icon with a tiny clock
-  }
-}
+  'custom-timeout': { label: 'If longer than %1', icon: bellOutlineIcon }, //Todo: change icon
```
We need to finish this, it is showing `%1`


![image](https://github.com/user-attachments/assets/28b96a27-d288-4c18-97db-8d626dea070b)


So what we could do is have a dropdown menu with all options show up on click, and then a nested menu with options for 1 minute, 5 minutes, 15 minutes. Maybe right-click should open a menu and left-click allow to cycle as currently is the case?

I mean something similar to the menu in code console in lab 4.4 [implemented here](https://github.com/jupyterlab/jupyterlab/blob/9749b07d54b4ac7622e617f1afdce770849c6f92/packages/console-extension/src/index.ts#L346-L354):

![image](https://github.com/user-attachments/assets/2cfb4d63-a545-427e-8520-aabfe0d7b713)

krassowski replied:
Also, currently this fails silently if custom timeout is not specified in settings, with the endpoint returning 400 and:

```
1 validation error for NotificationParams
timer
  Input should be an instance of Timer [type=is_instance_of, input_value=False, input_type=bool]
    For further information visit https://errors.pydantic.dev/2.10/v/is_instance_of
```

but user sees nothing.

Darshan808 replied:
Thanks, Added a error notification for it!

krassowski commented on src/index.ts:
```diff
@@ -2,150 +2,414 @@ import {
+          data: { [MIME_TYPE]: notificationData },
+        });
+        const renderer = rendermime.createRenderer(MIME_TYPE);
+        await renderer.renderModel(mimeModel);
+        console.log('Notification rendered successfully');
+        notification.notificationIssued = true;
+      } catch (err) {
+        console.error('Error rendering notification:', err);
+      }
```
This always errors out for me. What is the MIME  renderer needed for here?

```
index.ts:269 Error rendering notification: TypeError: Cannot read properties of undefined (reading 'application/desktop-notify+json')
    at OutputWidget.renderModel (mime.ts:72:48)
    at handleNotification (index.ts:265:24)
````

Darshan808 replied:
MIME renderer is needed for desktop notification. Caught this bug!
`mime.ts` was cloning an object by serializing and deserializing it, which caused the getters to be lost and threw an error. Fixed it!

krassowski commented on src/index.ts:
```diff
@@ -2,150 +2,414 @@ import {
+
+    // Settings management
+    const updateSettings = (settings: ISettingRegistry.ISettings): void => {
+      console.log('New Settings:', settings.composite);
```
Remainder to remove once you finished debugging

```suggestion
```

krassowski commented on schema/plugin.json:
```diff
@@ -12,6 +12,19 @@
+        "never",
+        "always",
+        "on-error",
+        "global-timeout",
```
Thinking more about it, maybe `global-timeout` and `always` should be combined and the default should be set to `always`? I suspect users would not want to have a notification of cells which take <5 seconds and they usually would want to have notifications on errors. I think maybe you raised it before but I missed this.

Darshan808 replied:
So to clarify, there shouldn’t be a `global-timeout` at all? And by default notifications should always be triggered if cell execution exceeds the threshold value? Also, what should the default threshold be set to?

krassowski replied:
Yes. Maybe we should remove `defaultMode` altogether and have three settings:
- [ ] notify on error
- [ ] notify after timeout
- default notification timeout [ number ]

I propose this because:
- if we name it "always" but then add threshold, it is a misnomer
- I somewhat worry that `mode`/`defaultMode` concept will make it harder to add further settings later, like different thresholds for different notification channels.

Then each cell has a mode override which is:
- "never" - do not notify about this cell
- "default" - follow the settings (as above)
- "on-error" - only notify on error, not on timeout
- "custom-timeout" - follow the settings, but use a different timeout

Now, I am not sure if this should be implemented as "modes", or reflect the settings structure proposed above, but the user-facing choices of per-cell options should be collapsed to the four options (where `custom-timeout` branches to a few options with different timeouts). What do you think?

> Also, what should the default threshold be set to?

I think something in range of 15 seconds, 30 seconds, or 1 minute?

Darshan808 replied:
> (where custom-timeout branches to a few options with different timeouts). What do you think?

Would it be better to let users set the timeout value via settings or limit it through a branching menu? I think it would be better to let them set it via settings.

krassowski replied:
> Would it be better to let users set the timeout value via settings or limit it through a branching menu? I think it would be better to let them set it via settings.

There would be two ways:
- global timeout - in settings
- per-cell timeout - branching menu (it could have "custom" entry that opens popup for more options)

Darshan808 replied:
Got it! Just two quick questions:

1. Does the default threshold also apply in 'on-error' mode, or does 'on-error' notify even when an error occurs immediately ?

2. Does the custom timeout mean:
   - "Notify me only if cell execution exceeds the timeout,"  
   - or "Notify me if cell execution exceeds the timeout, but also notify if it completes before the timeout"?

krassowski commented on README.md:
````diff
@@ -5,77 +5,75 @@
-```
+- `always`: Sends a notification every time a cell finishes executing.
+- `never`: Disables notifications for the cell.
+- `on-error`: Sends a notification only if the cell - execution fails with an error.
````
```suggestion
- `on-error`: Sends a notification only if the cell execution fails with an error.
```

krassowski reviewed (commented):
Great progress! Sharing feedback from a first pass on the code

Darshan808 committed:
modify-ui-and-default-notify

Darshan808 commented:
### Changes:
- [x] Updated UI to reflect a menu for notification modes instead of a click cycle.
- [x] Merged `global-timeout` and `always` modes into a new `default` mode. The `default` mode now notifies only if cell execution finishes and exceeds the threshold defined in settings (Default: 30 seconds).
- [x] Updated `README` to document the changes.
- [x] Added `slack_sdk` as a dependency in `pyproject.toml`.
- [x] Re-designed bell-clock SVG to match the others.
- [x] Fixed bugs, typos and applied feedbacks given above

### To do:
- [x] Modify tests (will do after finalizing features).

### Other-info:
> create a `jupyterlab_notify_config.json` configuration file under a path listed in `jupyter --paths` config section (e.g. 

Are these configurations automatically loaded and configured?
Because In code I'm doing it like this:
```python
    def _load_from_file(self, key):
        config_path = Path.home() / r".jupyter/jupyterlab_notify_config.json"
        try:
            with open(config_path, "r") as f:
                config = json.load(f)
                return config.get(key)
```
- I haven't tried setting up smtp locally. I just tested it using a dummy smtp that logs smtp messages in terminal. I haven't tried the link provided in read-me for smtp configuration

Darshan808 commented:
@krassowski 
I’ve been calculating the execution time from the timing **metadata** of cells by taking the difference between these two timestamps:
```python
        const startTime = timingData['shell.execute_reply.started'];
        const endTime = timingData['shell.execute_reply'] ?? timingData['execution_failed'];
```
However, when running tests, I noticed that execution timing metadata is missing from the cells:
![Screenshot from 2025-03-21 23-39-10](https://github.com/user-attachments/assets/078f1363-d19f-4781-963a-a8b7abe04e65)

Isn't JupyterLab supposed to provide this metadata by default? Any idea why it might be missing in the test environment?

Darshan808 commented:
> Isn't JupyterLab supposed to provide this metadata by default?

Oh, seems like `record-timing` setting is by default set to false. Handling this!
Is it okay to depend upon cell metadata for execution timing data ? Or should I look for other approach ? It is working fine.

Darshan808 committed:
handle-record-timing-settings

Darshan808 commented:
> Oh, seems like record-timing setting is by default set to false. Handling this!

Handled this by displaying a dialog box when a user opens a new notebook if record_timing is not enabled. Any feedback on this approach?

krassowski commented:
> > Oh, seems like record-timing setting is by default set to false. Handling this!
>
> Handled this by displaying a dialog box when a user opens a new notebook if record_timing is not enabled. Any feedback on this approach?

We could auto-enable it, this is what `jupyterlab-execute-time` [does](https://github.com/deshaw/jupyterlab-execute-time/blob/d9011e9ed0ce32a8da811519d83a0ed4d6605553/src/index.ts#L54-L66) and I think is a better UX (probably should be mentioned in the readme).

Darshan808 committed:
modify-workflow

krassowski commented:
> create a jupyterlab_notify_config.json configuration file under a path listed in jupyter --paths config section (e.g.
> 
> Are these configurations automatically loaded and configured?

These should be auto-loaded onto traitlets for `ExtensionApp` subclasses based on name. Have a look around at how https://github.com/nebari-dev/jupyterlab-gallery has `jupyter_gallery_config.json` auto loaded. Note that its name is [`gallery`](https://github.com/nebari-dev/jupyterlab-gallery/blob/b2cad9c33d2e94e57af50b67d522ffe46bc2afe7/jupyterlab_gallery/app.py#L8) so for an extension named `jupyterlab_notify` it would load it up from `jupyter_jupyterlab_notify_config.json`. Also see https://jupyter-server.readthedocs.io/en/latest/developers/extensions.html#server-extensions

krassowski commented:
So the current name is `jupyter_notify_v2`, since it will work in both Notebook and JupyterLab I would just change it to `notify`. Another thing is that the passed traitlets will only be applied to traits defined on:
- the `ExtensionApp` class
- on of its registered handlers
- one of the classes defined on `classes` attribute, see [here](https://github.com/jupyter-server/jupyter_server/blob/2cf8ac5b2bc2aa6aeb14a6e3f56230c160604b4b/jupyter_server/extension/application.py#L208-L211)

Darshan808 committed:
force-record-timing

Darshan808 committed:
fix-workflow

Darshan808 committed:
reuse-existing-server

Darshan808 committed:
make-config-configurable

Darshan808 committed:
update-readme

Darshan808 committed:
modify-test

Darshan808 committed:
remove-bitwise-OR-use-Optional

Darshan808 commented:
@krassowski 
Normally, everything works fine. However, when installing with `pip install ".[server-side-execution]"`, only `jupyter_server_nbmodel` gets installed. As a result, outputs and metadata aren't sent to the frontend. Since notifications depend on execution timing metadata, this causes notification delivery to fail.  

In the README of `jupyter-server-nbmodel`, I found: 
> Or with recommendations:
> ```bash
> pip install jupyter_server_nbmodel[rtc]
> ```

However, I don’t see any optional dependency group named `rtc` in its `pyproject.toml`.  
https://github.com/datalayer/jupyter-server-nbmodel/blob/efc2f849e0f7dd388ac93d0c1b1fca3991885707/pyproject.toml#L31C1-L42C25

What things are required to push outputs and metadata to the client when using `jupyter-server-nbmodel`?

krassowski commented:
> However, I don’t see any optional dependency group named rtc in its pyproject.toml.
https://github.com/datalayer/jupyter-server-nbmodel/blob/efc2f849e0f7dd388ac93d0c1b1fca3991885707/pyproject.toml#L31C1-L42C25
> 
> What things are required to push outputs and metadata to the client when using jupyter-server-nbmodel?

We should open a PR to fix it. Here we would use a part of the `lab` group, as in an earlier comment:

> I think this also requires "jupyter-docprovider>=1.0.0b1", "jupyter-server-ydoc>=1.0.0b1" as in the [lab] optional group of jupyter-server-nbmodel ([here](https://github.com/datalayer/jupyter-server-nbmodel/blob/efc2f849e0f7dd388ac93d0c1b1fca3991885707/pyproject.toml#L32)).

Darshan808 commented:
Everything is working fine. Need some feedback with these small issues:

i. Make the toolbar button tooltip dynamic (e.g., `mode (click to change)`). I changed the button icon in a bit tricky way but couldn't update the tooltip.

ii. Show the selected mode. Other menus use a checkmark to indicate the selected item, but since we're using icons, it's tricky to apply the same approach.

krassowski commented:
Binder: https://mybinder.org/v2/gh/Darshan808/jupyterlab-notify.git/v2-test?urlpath=lab%2Ftree%2Fnotebooks%2Findex.ipynb

Some improvements to consider:
- [x] Fix the check links job
- [x] Fix binder to use an existing notebook (it points to `notebooks/index.ipynb` but there is only `notebooks/Notify.ipynb`)
- [x] Make binder install the `[server-side-execution]` dependencies
- [x] Update `notebooks/Notify.ipynb`
- [x] Improve alignment of icons in menu
     ![image](https://github.com/user-attachments/assets/4191e055-559b-4939-a990-77a402866565)
- [x] Maybe rename settings section from "Notify" to "Execution Notifications" to disambiguate it from built-in "Notifications" settings?
- [x] Change "1 hr" to "1 hour"?

Darshan808 committed:
add-demo-notebook-and-other-fixes

Darshan808 commented:
I’ve fixed all the issues, but I’m unsure about the broken links in the changelog. Any insights on how to handle them?

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-lint-and-tests

Darshan808 committed:
fix-links

Darshan808 commented:
There are only four tags in [jupyterlab-notify](https://github.com/deshaw/jupyterlab-notify/tags). I updated the links in the changelog based on the changes introduced in each tag and documented in the changelog.  

I'm not entirely sure if this fix is correct.

Darshan808 committed:
add-docs-images

Darshan808 commented on README.md:
```diff
@@ -56,29 +56,28 @@ These settings allow for customization, such as using a custom SMTP server or ch
 You can control when notifications are sent by setting a mode for each cell. Modes can be configured through the JupyterLab interface by clicking on the bell icon in the cell toolbar.
 
-![image](https://github.com/user-attachments/assets/52c0f81a-fdf3-4840-b168-4c067b2b1251)
+![image](https://github.com/darshan808/jupyterlab-notify/blob/v2-test/docs/celltoolbar-menu-screenshot.png?raw=true)
```
I've temporarily linked to my forked branch to display the images. I'll update the links once the changes are finalized for merging.

Darshan808 committed:
fix-links

Darshan808 committed:
enhance-readme

Darshan808 committed:
remove-custom-timout-from-settings

Darshan808 commented:
Made the `Notify.ipynb` demo more interactive and updated the Binder link to point to it. Here's the updated link:  
[binder](https://hub.2i2c.mybinder.org/user/darshan808-jupyterlab-notify-zqd6vz0y/lab/workspaces/auto-6/tree/notebooks/Notify.ipynb)

krassowski commented on src/index.ts:
```diff
@@ -2,152 +2,581 @@ import {
+        label: 'Help',
+        callback: () =>
+          showErrorMessage(`${service} Not Configured`, {
+            message: `Add a ${service.toLowerCase()} configuration to directory listed under the config section of jupyter --paths (e.g., ~/.jupyter/jupyter_notify_config.json) to enable ${service.toLowerCase()} notifications. Example: \n{\n  "${configKey}": "${example}"}-config"\n}. If you've already configured it, there might be an issue. Please check the terminal for errors and review your set ...
```
It looks like that part is spurious?

```suggestion
            message: `Add a ${service.toLowerCase()} configuration to directory listed under the config section of jupyter --paths (e.g., ~/.jupyter/jupyter_notify_config.json) to enable ${service.toLowerCase()} notifications. Example: \n{\n  "${configKey}": "${example}"}. If you've already configured it, there might be an issue. Please check the terminal for errors and review your setup.`,
```

![image](https://github.com/user-attachments/assets/4aac4a92-4035-4d1d-a3ed-20a0b2b077f2)

krassowski replied:
The same dialog is displayed when email is not configured and when SMPT server is not running, which for me shows in jupyter-server logs: `SMTP Configuration Error: Failed to instantiate SMTP: [Errno 111] Connection refused`. Can we add another variable in initial response (in addition to `email_configured`) to indicate if the server is available and display another error if it is not?

Darshan808 committed:
add-error-msg-for-smtp-server

Darshan808 committed:
remove-unwanted-files-from-git

Darshan808 committed:
add-batch-notify

Darshan808 committed:
restore-magics

Darshan808 committed:
add-click-to-cell-on-notification

Darshan808 committed:
bring-settings-to-notebook-level

Darshan808 committed:
add-tooltips-and-notebook-toolbar-menu

krassowski commented on src/utils.ts:
```diff
@@ -0,0 +1,282 @@
+    const node = document.createElement('div');
+    node.className = 'jp-notify-time-input-dialog-body';
+
+    const label =
+      options.label || 'Default: seconds, +m for minutes, +h for hours:';
+
+    node.innerHTML = `
+      <label class="jp-notify-time-input-label">${label}</label>
+      <div class="jp-notify-time-input-container">
+        <input
+          type="number"
+          class="jp-notify-time-input-field"
+          placeholder="${options.placeholder || '30'}"
+          min="0"
+          step="any"
+        />
+        <select class="jp-notify-time-unit-select">
+          <option value="${TimeUnit.SECONDS}">Seconds (s)</option>
+          <option value="${TimeUnit.MINUTES}">Minutes (m)</option>
+          <option value="${TimeUnit.HOURS}">Hours (h)</option>
+        </select>
+      </div>
+      <div class="jp-notify-time-input-error" role="alert">Please enter a valid positive number</div>
+    `;
```
I would be worried about a potential for unsanitised injection here, could this be a react-based widget instead? If not, let's set the placeholer and label separately by querying relevant nodes.

krassowski commented:
It looks the build fails on linting.

Darshan808 committed:
modify to react widget

Darshan808 committed:
fix-tests

Darshan808 commented:
UI test is failing due to first cell not being clickable due to toolbar element intercepting pointer events.
<img width="878" height="163" alt="image" src="https://github.com/user-attachments/assets/c773b334-9b0d-41f1-8fef-a7241ed68b20" />
This might be due to the notebook toolbar being overcrowded and items not being hidden properly. This might be a bug in JupyterLab's toolbar.

Darshan808 committed:
fix-test

Darshan808 committed:
fix-test

Darshan808 committed:
improve-icons

Darshan808 committed:
handle-null-thresholds

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-binder-and-some-code

Darshan808 committed:
fix-bug-with-custom-timeout-always-triggering-notification

Darshan808 committed:
notify-on-kernel-death-and-autorestarts

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-test

Darshan808 committed:
fix-test

Darshan808 committed:
add-optional-slack-group

Darshan808 committed:
add-notebook-name-in-notification

Darshan808 committed:
lint code

Darshan808 committed:
persist-threshold

Darshan808 committed:
use-unique-keys-for-threshold-and-timeout

Darshan808 committed:
use-checkboxes-to-replace-all-thresholds

krassowski commented on src/index.ts:
```diff
@@ -1162,6 +1276,95 @@ const plugin: JupyterFrontEndPlugin<void> = {
+                          .replace(/\s+/g, '-')
+                          .replace(/\(\d+[smh]\)$/, '')
+                          .trim();
+                      }
```
Ouch. Maybe we could subclass `MenuSvg` to influence the renderer?

Darshan808 committed:
make-slack-and-email-messages-better

Darshan808 committed:
send-timeout-notifications-only-once

Darshan808 committed:
subclass-menu

Darshan808 committed:
seperate-tokens

Darshan808 committed:
clean code

Darshan808 committed:
fix-payload

Darshan808 committed:
do-not-display-exec-count-on-sse

Darshan808 committed:
run-lint

Darshan808 committed:
use-optional-instead

Darshan808 committed:
add-notebook-tests

Darshan808 committed:
UX: add-checkmarks

Darshan808 commented:
The test fails because in small screen when we open sidebars, the notebook toolbar gets cluttered and `Python 3 (ipykernel)` comes down and overlays the 1st cell in the UI.

<img width="994" height="299" alt="Screenshot 2026-03-06 at 14 19 22" src="https://github.com/user-attachments/assets/d06d6c0d-417c-4992-9d51-0eb046a9b1be" />

```bash
  91 × waiting for element to be visible, enabled and stable
       - element is visible, enabled and stable
       - scrolling into view if needed
       - done scrolling
       - <span class="jp-ToolbarButtonComponent-label">Python 3 (ipykernel)</span> from <jp-toolbar role="toolbar" orientation="horizontal" aria-orientation="horizontal" aria-label="main area toolbar" class="lm-Widget jp-Toolbar jp-NotebookPanel-toolbar">…</jp-toolbar> subtree intercepts pointer events
     - retrying click action
       - waiting 500ms


  76 |
  77 |   // Add a second cell and set its threshold
> 78 |   await page.notebook.addCell('code', '# cell 2');
     |   ^
```

I have encountered this before. Let's see what we can do to solve this.

Darshan808 committed:
try-fix-test

Darshan808 committed:
try-fix-test-again

Darshan808 committed:
add-missing-wait

Darshan808 commented:
Closing the sidebars before adding the new cell did the job!

krassowski commented on jupyterlab_notify/handlers.py:
```diff
@@ -0,0 +1,153 @@
+        )
+
+        # If a timeout threshold is configured, schedule a timer to trigger notification.
+        if params.mode in ("custom-timeout"):
```
Should it be

```suggestion
        if params.mode == "custom-timeout":
```

krassowski commented on src/menuTooltip.ts:
```diff
@@ -0,0 +1,47 @@
+          li.setAttribute('title', tooltip);
+        }
+
+        const checked = args?.checked as boolean | undefined;
+        const iconNode = li.querySelector('.lm-Menu-itemShortcut');
+        if (iconNode) {
+          if (checked) {
+            iconNode.innerHTML = checkIcon.svgstr;
+          } else {
+            iconNode.innerHTML = '';
```
Can you improve the alignment for the check icon? It seems vertically off center right now:

<img width="443" height="252" alt="Image" src="https://github.com/user-attachments/assets/bc3d5853-088c-4c87-ac52-e11ba30a13e3" />

krassowski commented on schema/plugin.json:
```diff
@@ -1,22 +1,52 @@
+        "name": "setNotebookDefaultThreshold",
+        "command": "notify:set-notebook-default-threshold",
+        "rank": 902
+      }
```
Are these two buttons redundant with the menu options? Should we drop them?

<img width="359" height="226" alt="Image" src="https://github.com/user-attachments/assets/5f7dbebd-b595-489e-afb6-5ddd8c573b90" />

Darshan808 replied:
Yes, I agree. The notebook toolbar already seems quite cluttered.

krassowski commented on src/mime.ts:
```diff
@@ -87,28 +115,103 @@ class OutputWidget extends Widget implements IRenderMime.IRenderer {
+          behavior: 'smooth',
+          block: 'center',
+          inline: 'nearest',
+        });
```
Changing `activeCellIndex` will already scroll the cell into view so smooth scrolling would not have an effect I believe.

In `full` windowing mode smooth scrolling this won't work.

Maybe we should use the public API for scrolling to cells. Maybe that API should support smooth scrolling.

Darshan808 replied:
Yes. I replaced it with `Notebook.scrollToCell`. But it doesn't seem to support smooth scrolling. Ig we could open an issue on jupyterlab.

krassowski commented on src/mime.ts:
```diff
@@ -87,28 +115,103 @@ class OutputWidget extends Widget implements IRenderMime.IRenderer {
+
+  private navigateToCellInNotebook(notebook: any, cellId: string): boolean {
+    const cells = notebook.content.widgets;
+    const cellIndex = cells.findIndex((cell: any) => cell.model.id === cellId);
```
Can you add an eslint rule to forbid `any` outside of tests?

krassowski commented on src/mime.ts:
```diff
@@ -87,28 +115,103 @@ class OutputWidget extends Widget implements IRenderMime.IRenderer {
+    const cellNode = cell.node;
+    const originalBackground = cellNode.style.backgroundColor;
+    const originalTransition = cellNode.style.transition;
+
+    // Add highlight with animation
+    cellNode.style.transition = 'background-color 0.3s ease';
+    cellNode.style.backgroundColor = '#fff3cd';
+
+    setTimeout(() => {
+      cellNode.style.backgroundColor = originalBackground;
+
+      setTimeout(() => {
+        cellNode.style.transition = originalTransition;
+      }, 300);
```
When the same cell notifies twice and user clicks on the notification twice before the background has cleared, they are left with the background stuck, because the `originalBackground` gets overwritten by what was set earlier:

<img width="375" height="54" alt="Image" src="https://github.com/user-attachments/assets/6baade8f-21b0-4271-997b-a5f8c8fc92c6" />

I think just setting both to `''` would be good enough.

krassowski commented on src/mime.ts:
```diff
@@ -87,28 +115,103 @@ class OutputWidget extends Widget implements IRenderMime.IRenderer {
+
+    // Add highlight with animation
+    cellNode.style.transition = 'background-color 0.3s ease';
+    cellNode.style.backgroundColor = '#fff3cd';
```
This does not look good in the dark theme:

<img width="375" height="54" alt="Image" src="https://github.com/user-attachments/assets/3e61289c-79e7-4c92-968f-c4b1d67390c6" />

I think you would need to add a new CSS variable for it.

Darshan808 replied:
What do you think about this one?
<img width="581" height="65" alt="Screenshot 2026-03-07 at 11 25 16" src="https://github.com/user-attachments/assets/d8a84f99-5995-47d5-90fb-dfd4eb0316d1" />

With error:
<img width="921" height="175" alt="Screenshot 2026-03-07 at 11 24 26" src="https://github.com/user-attachments/assets/06e273f1-50fc-4293-b7bf-df66fff2510d" />

krassowski replied:
I think it's fine, its just a short flash so contrast does not need to be perfect.

krassowski commented on jupyterlab_notify/extension.py:
```diff
@@ -0,0 +1,224 @@
+
+        # Build formatted message with status, cell info, and details
+        cell_info = f"Cell: {params.execution_count}" if params.execution_count is not None else f"Cell id: {params.cell_id}"
+        print(f"\nParams: {params}")
```
Debugging leftover?

```suggestion
```

Darshan808 committed:
forbid-explicit-any

Darshan808 committed:
remove-debugging-leftovers

Darshan808 committed:
apply-feedbacks

Darshan808 committed:
remove-toolbar-buttons

Darshan808 committed:
use-menu-in-tests-to-change-threshold

krassowski commented:
Given:

<img width="788" height="246" alt="image" src="https://github.com/user-attachments/assets/ad2c7746-c046-47e8-bbf6-7c3b10db6381" />

Notification from custom timeout on `sleep(4)`:

<img width="528" height="112" alt="image" src="https://github.com/user-attachments/assets/cfda8b8d-e363-42ce-ba86-948963ed984e" />

Notification from "on error" on `raise`:

<img width="528" height="112" alt="image" src="https://github.com/user-attachments/assets/3dd3e9cc-fcba-425e-b290-861f76218f94" />

1. Is this expected that the former has a blank line rather than Cell: 10?
2. Should the latter also show the error message (when easily available)
3. Should we use [10] and [11] to keep it short?

krassowski commented:
I think we should have a setting for "Always notify on error" which should be on by default. Then "On error" mode would mean "_only_ on error" when that setting is enabled.

krassowski commented on schema/plugin.json:
```diff
@@ -1,22 +1,42 @@
+      "title": "Default mode",
+      "description": "The default mode for notification",
+      "type": "string",
+      "enum": ["default", "never", "on-error", "custom-timeout"],
```
We should separate code names from user-facing strings by using `oneOf` as [in here](https://github.com/jupyterlab/jupyterlab/blob/badbcb1797d40619310ca8a057367bce55105548/packages/apputils-extension/schema/kernels-settings.json#L20-L22)


```suggestion
      "oneOf": [
        { "const": "default", "title": "Default" },
        { "const": "never", "title": "Never" },
        { "const": "on-error", "title": "On Error" },
        { "const": "custom-timeout", "title": "Custom timeout" }
        "default", "", "on-error", ""
      ],
```

krassowski commented on schema/plugin.json:
```diff
@@ -1,22 +1,42 @@
+    },
+    "customTimeout": {
+      "title": "Timeout value for timeout notifications",
+      "description": "Default timeout value for custom-timeout notification (unit in seconds)",
```
```suggestion
      "description": "Default timeout value for custom timeout notification (unit in seconds)",
```

krassowski commented on docs/celltoolbar-menu-screenshot.png:
It feels somewhat odd that error in "On error" is lowercase when everything else is using Title Case. This screenshot is also slightly out of date.

<img width="403" height="224" alt="Image" src="https://github.com/user-attachments/assets/ade29c3a-55ae-4b30-b02b-95ecea50e831" />

krassowski commented on src/index.ts:
```diff
@@ -1,153 +1,1255 @@
+
+      // Handle case when threshold isn't exceeded in default mode
+      if (payload.mode === 'default') {
+        const timingData: IExecutionTimingMetadata =
+          cell.getMetadata('execution');
```
On binder the setup sometimes fails due to some ydoc extension loading issue:

<img width="1058" height="435" alt="image" src="https://github.com/user-attachments/assets/4751e408-d802-465d-a6b4-d4c4cfb21d06" />

This leads to a failure:

<img width="848" height="817" alt="image" src="https://github.com/user-attachments/assets/d5224d77-57d8-4973-928b-6b571a7565a3" />

Instead of casting `getMetadata()` result we could check if it is `undefined` or not, but I am not sure what is the appropriate way to handle such case. I am mor worried that the activation failure occurs in the first place.

Darshan808 replied:
> On binder the setup sometimes fails due to some ydoc extension loading issue:

Yes, It sometime happens on my machine locally as well.
If I reload the browser the error disappears.

> but I am not sure what is the appropriate way to handle such case.

For now I think we should avoid sending notification if timing metadata isn't present.

krassowski commented on notebooks/Notify.ipynb:
I think metadata in this example notebook needs updating too.

krassowski commented on jupyterlab_notify/magics.py:
```diff
@@ -78,7 +75,7 @@ class NotifyCellCompletionMagics(Magics):
         super(NotifyCellCompletionMagics, self).__init__(shell)
         self.smtp_instance = None
-        self._setup_smtp_instance()
+        # self._setup_smtp_instance()
```
Is this commented out intentionally?

Darshan808 replied:
I don't think so. It might be just testing leftover. Thanks.

Darshan808 committed:
run-precommit

Darshan808 committed:
address-copilot-feedbacks

Darshan808 commented:
> 1 Is this expected that the former has a blank line rather than Cell: 10?

Yes. When cell gets scheduled for execution, it looses its executionCount number. So, I think even if we somehow show its previous executionCount, the user won't be able to see it in UI.

> 2. Should the latter also show the error message (when easily available)

Yes, I think should include it.

> 3. Should we use [10] and [11] to keep it short?

So you mean like this?
```
[test_notebook] Cell execution successful
[10]
```
OR: 
```
test_notebook[10]: Cell execution successful
```
And in error:
```
[test_notebook] Cell execution failed
[10]: ZeroDivisionError: division by zero
```
OR
```
test_notebook[10]: Cell execution failed
ZeroDivisionError: division by zero
```

For long error messages, I think we could truncate error message and show ellipsis `[10]: ZeroDivisionError: division by z...`

Darshan808 committed:
address-feedbacks

Darshan808 committed:
Fix-lint

krassowski commented:
Both look good, though two suggesions:
- if we are keeping the second line maybe populate it with timing info
- I think users don't see notebook name in `[]` elsewhere so maybe we should avoid it

```
test_notebook: Cell execution successful
Cell [10] completed in X seconds
```

```
test_notebook: Cell execution failed
Cell [10]: ZeroDivisionError: division by zero
```

As in debugger we show:

<img width="288" height="200" alt="image" src="https://github.com/user-attachments/assets/44d97531-8144-41ef-9245-7a5a508948ff" />

krassowski commented:
> Is this expected that the former has a blank line rather than Cell: 10?

So basically we would need to delay the notification slightly, because unless the cell fails due to kernel restart I would expect the execution counter to be updated eventually (as it shows up in the UI). I think that's ok to delay it as it is milliseconds difference?

Darshan808 commented:
> So basically we would need to delay the notification slightly,

Yes. But cell number doesn't appear in notification only in case of `custom timeout`. Because it notifies mid execution as soon as it timeouts when cell doesn't have execution count. So, even if we delay, we won't know when it might finish.
In all other types, we do include cell number.

Darshan808 committed:
improve-notification

krassowski commented:
Ah, ok that make's sense.

krassowski commented on src/index.ts:
```diff
@@ -1,153 +1,1269 @@
-}
+    label: 'Custom Timeout',
+    icon: bellClockIcon,
+    info: 'Notify after a custom timeout, regardless of execution result.',
```
Is my suggestion correct?

```suggestion
    info: 'Notify if a cell is still running after a set timeout.',
```

Darshan808 committed:
add-alwaysNotifyOnError

Darshan808 committed:
update-notebook

Darshan808 committed:
modify-docs-snapshots

Darshan808 committed:
improve-description

Darshan808 commented:
> I think we should have a setting for "Always notify on error" which should be on by default. Then "On error" mode would mean "only on error" when that setting is enabled.

Yes this totally makes sense. I just added it.

Darshan808 committed:
minor-comment

krassowski reviewed (approved):
Thank you @Darshan808 - let's release a new demo version 🎉

Darshan808 commented:
> let's release a new demo version 🎉

Just released it!
`pip install jupyterlab-notify-v2-demo==2.0.7`

krassowski commented:
Thank you! One thing I spotted is that:

```
import os, signal
os.kill(os.getpid(), signal.SIGKILL)
```

notifies when using "On Error" mode but does not when using the default mode.

Darshan808 committed:
fix-kernel-death-notification

Darshan808 commented:
> notifies when using "On Error" mode but does not when using the default mode.

Just fixed it. Should I go ahead and release another demo version which includes the fix. Or are we good with the current one?

krassowski commented:
If that's less than 15 minutes then I would say release?

Darshan808 commented:
Done!
`pip install jupyterlab-notify-v2-demo==2.0.8`

Darshan808 committed:
apply-suggestions

Darshan808 committed:
add-cell-number-for-timeout

Darshan808 committed:
run-lint

Darshan808 committed:
add-tests

krassowski commented:
Jut highlighting:

```
[warn] ui-tests/tests/test_cell.spec.ts
[warn] Code style issues found in the above file. Forgot to run Prettier?
```

Darshan808 committed:
lint-code
